File: app/src/main/java/com/github/goldy1992/mp3player/client/IntentMapper.java
package com.github.goldy1992.mp3player.client;

import android.content.Context;
import android.content.Intent;

import androidx.annotation.Nullable;

import com.github.goldy1992.mp3player.client.activities.FolderActivityInjector;
import com.github.goldy1992.mp3player.client.activities.MediaActivityCompat;
import com.github.goldy1992.mp3player.client.activities.MediaPlayerActivityInjector;
import com.github.goldy1992.mp3player.commons.MediaItemType;
import com.github.goldy1992.mp3player.dagger.scopes.ComponentScope;

import java.util.EnumMap;
import java.util.Map;

import javax.inject.Inject;

@ComponentScope
public class IntentMapper {

    private final Map<MediaItemType, Class<? extends MediaActivityCompat>> categoryToActivityMap = new EnumMap<>(MediaItemType.class);

    private final Context context;

    @Inject
    public IntentMapper(Context context) {
        this.context = context;
        init();
    }

    private void init() {
        // TODO: change this code to accommodate test implementations. Intents should be made in Dagger
        categoryToActivityMap.put(MediaItemType.FOLDER, MediaPlayerActivityInjector.class);
        categoryToActivityMap.put(MediaItemType.SONGS, MediaPlayerActivityInjector.class);
        categoryToActivityMap.put(MediaItemType.FOLDERS, FolderActivityInjector.class);

    }

    @Nullable
    public Intent getIntent(MediaItemType mediaItemType) {
        Class<? extends MediaActivityCompat> clazz = categoryToActivityMap.get(mediaItemType);
        if (null != clazz) {
            return new Intent(context, clazz);
        }
        return null;

    }

}

File: client/src/main/java/com/github/goldy1992/mp3player/client/IntentMapper.kt
package com.github.goldy1992.mp3player.client

import android.content.Context
import android.content.Intent
import com.github.goldy1992.mp3player.commons.ComponentClassMapper
import com.github.goldy1992.mp3player.commons.MediaItemType
import com.github.goldy1992.mp3player.commons.dagger.scopes.ComponentScope
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.android.scopes.ActivityScoped
import java.util.*
import javax.inject.Inject

@ActivityScoped
class IntentMapper
    @Inject
    constructor(@ApplicationContext private val context: Context,
                private val componentClassMapper: ComponentClassMapper) {
    private val categoryToActivityMap: MutableMap<MediaItemType,
            Class<*>> = EnumMap(MediaItemType::class.java)
    init {
        categoryToActivityMap[MediaItemType.FOLDER] = componentClassMapper.mediaPlayerActivity!!
        categoryToActivityMap[MediaItemType.SONGS] = componentClassMapper.mediaPlayerActivity!!
        categoryToActivityMap[MediaItemType.FOLDERS] = componentClassMapper.folderActivity!!
    }

    fun getIntent(mediaItemType: MediaItemType?): Intent? {
        val clazz = categoryToActivityMap[mediaItemType]
        return clazz?.let { Intent(context, it) }
    }

}